lib/sort.dx

```diff
@@ -11,32 +11,33 @@ However, if breaks the list in half recursively, it'll be doing parallel mergeso
 Currently it'll do the quadratic time version.
 
 def concatTable {a b v} (leftin: a=>v) (rightin: b=>v) : ((a|b)=>v) =
-  -- Note: It'd be nicer to use (a | b)=>v but it's not currently supported.
   for idx. case idx of
     Left i  -> leftin.i
     Right i -> rightin.i
 
 def mergeSortedTables {a m n} [Ord a] (xs:m=>a) (ys:n=>a) : ((m|n)=>a) =
   -- Possible improvements:
   --  1) Using a SortedTable type.
-  --  2) Avoid initializing the return array.
-  init = concatTable xs ys
+  --  2) Avoid pontlessly initializing the return array.
+  init = concatTable xs ys  -- Initialize array of correct size.
   yieldState init \buf.
     withState (0, 0) \countrefs.
-      for i.
+      for i:(m|n).
         (cur_x, cur_y) = get countrefs
-        noYsLeft = cur_y >= size n
-        stillXsLeft = cur_x < size m
-        cur_x_at_n = (unsafeFromOrdinal _ cur_x)
-        cur_y_at_n = (unsafeFromOrdinal _ cur_y)
-        xIsLess = xs.cur_x_at_n < ys.cur_y_at_n
-        if noYsLeft || (stillXsLeft && xIsLess)
+        if cur_y >= size n  -- no ys left
           then
             countrefs := (cur_x + 1, cur_y)
-            buf!i := xs.cur_x_at_n
+            buf!i := xs.(unsafeFromOrdinal _ cur_x)
           else
-            countrefs := (cur_x, cur_y + 1)
-            buf!i := ys.cur_y_at_n
+            if cur_x < size m -- still xs left
+              then 
+                if xs.(unsafeFromOrdinal _ cur_x) <= ys.(unsafeFromOrdinal _ cur_y)
+                  then
+                    countrefs := (cur_x + 1, cur_y)
+                    buf!i := xs.(unsafeFromOrdinal _ cur_x)
+                  else
+                    countrefs := (cur_x, cur_y + 1)
+                    buf!i := ys.(unsafeFromOrdinal _ cur_y)
```

Comment:
Would this be cleaner if you used a little helper such as this:
```
def next {n} [Ix n] (i:n) : Maybe n =
  next_ord = ordinal i + 1
  case next_ord < size n of
    True -> Just $ unsafeFromOrdinal _ next_ord
    False -> Nothing
```

Reply:
That's really nice, and it does get rid of all the unsafe casting.  But it gave me a `leaked local variable` error that I couldn't make heads of tails of, so I'll leave it for now.